### PR TITLE
bpftimetool: fixes unnecessary report of shm removed when there's nothing to remove

### LINKED ISSUE #169: [BUG] bpftimetool should not report shm removed when there is nothing to remove
Currently, running `bpftimetool` multiple times always gives the message:

```
[agentzh@fed32-dev ~/bpftime 134 (master)]$ bpftimetool remove
[2024-01-27 12:57:56.552] [info] [bpftime_shm_internal.cpp:46] Global shm removed
[agentzh@fed32-dev ~/bpftime 0 (master)]$ sudo bpftimetool remove
[2024-01-27 12:57:58.261] [info] [bpftime_shm_internal.cpp:46] Global shm removed
[agentzh@fed32-dev ~/bpftime 0 (master)]$ sudo bpftimetool remove
[2024-01-27 12:58:01.619] [info] [bpftime_shm_internal.cpp:46] Global shm removed
```

It's confusing to see it reports "Global shm removed" when there is actually nothing to remove...

### PR DESCRIPTION
**Pull request by aneeshdamle11**
## Description

bpftimetool reports "Global shm removed" even if there is nothing to remove. Fixed here.

Fixes #169

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

- [ ] Test A
- [ ] Test B

**Test Configuration**:

- Firmware version:
- Hardware:
- Toolchain:
- SDK:

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: runtime/src/bpftime_shm_internal.cpp [hunk 1]
```diff
@@ -41,9 +41,11 @@ extern "C" void bpftime_destroy_global_shm()
 extern "C" void bpftime_remove_global_shm()
 {
 	using namespace bpftime;
-	boost::interprocess::shared_memory_object::remove(
-		get_global_shm_name());
-	SPDLOG_INFO("Global shm removed");
+	const char *shm_name = get_global_shm_name();
+	boost::interprocess::shared_memory_object::remove(shm_name);
+	if (shm_name != DEFAULT_GLOBAL_SHM_NAME) {
```
Comment (Officeyutong) [inline runtime/src/bpftime_shm_internal.cpp:46]: Why compare c-string in this way?

Comment (aneeshdamle11) [reply runtime/src/bpftime_shm_internal.cpp:46]: I'll correct it in the next commit with strcmp

**Comment by yunwei37** [comment]
See https://www.boost.org/doc/libs/1_80_0/doc/html/interprocess/sharedmemorybetweenprocesses.html#interprocess.sharedmemorybetweenprocesses.sharedmemory.removing

Does that mean you should compare the `shm_name` pointer with nullptr? What is the `shm_name` pointer after the sharedmemory is successfully removed?

**Comment by aneeshdamle11** [comment]
I looked in the code further and pored over my approach. bpftime_remove_global_shm_name() should remove the shared memory segment from its system.
It gets the name of the global shared memory from get_global_shm() which fetches the value of an environment variable BPFTIME_GLOBAL_SHM_NAME.
I am a bit confused on how to test it properly. Could you help with any inputs?

**Comment by aneeshdamle11** [comment]
According to the referred doc: https://www.boost.org/doc/libs/1_80_0/doc/html/interprocess/sharedmemorybetweenprocesses.html#interprocess.sharedmemorybetweenprocesses.sharedmemory.removing, I changed the remove function to check it's return value. I realised I didn't need the shm_name variable to do this.

I tested the bpftimetool assigning a shared memory using the load function. Then I removed it twice, the second time showing no output.
image

Does this test work? I am a bit new code, but tried my best understanding this part. Please provide any inputs if my understanding needs to be improved.

**Review by Officeyutong — APPROVED** [review]
Thanks for your work!

**Comment by Officeyutong** [comment]
> According to the referred doc: https://www.boost.org/doc/libs/1_80_0/doc/html/interprocess/sharedmemorybetweenprocesses.html#interprocess.sharedmemorybetweenprocesses.sharedmemory.removing, I changed the remove function to check it's return value. I realised I didn't need the shm_name variable to do this.
> 
> I tested the bpftimetool assigning a shared memory using the load function. Then I removed it twice, the second time showing no output. image
> 
> Does this test work? I am a bit new code, but tried my best understanding this part. Please provide any inputs if my understanding needs to be improved.

I think it's correst. I'll merge this PR once the CI passed

**Comment by aneeshdamle11** [comment]
Glad I could help! And thank you for giving me this opportunity to contribute!

**Pull request merged**